packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,90 @@
+    const associationKey = this.getAssociationKey(scope, portfolio, product);
+    if (this.checkConstraintOrAssocationExists(associationKey, Constraints.ASSOCIATION)) {
+      const association = new CfnPortfolioProductAssociation(scope, `PortfolioProductAssociation${associationKey}`, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+
+      const associationValue = new Map<Constraints, AssociationOrConstraint>();
+      associationValue.set(Constraints.ASSOCIATION, association);
+      this.associationMap.set(associationKey, associationValue);
+    }
+  }
+
+  public static allowTagUpdates(
+    scope: cdk.Resource, portfolio: IPortfolio, product: IProduct,
+    options: TagUpdateConstraintOptions,
+  ): void {
+    InputValidator.validateLength(this.prettyPrintAssociation(portfolio, product), 'description', 0, 2000, options.description);
+    const associationKey = this.ensureProductIsAssociatedWithPortfolio(scope, portfolio, product);
+    if (this.checkConstraintOrAssocationExists(associationKey, Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.messageLanguage,
+        description: options.description,
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+        tagUpdateOnProvisionedProduct: options.allowUpdatingProvisionedProductTags === false ? 'NOT_ALLOWED' : 'ALLOWED',
+      });
+
+      // Add dependsOn to force proper order in deployment.
+      constraint.addDependsOn(this.associationMap.get(associationKey)!.get(Constraints.ASSOCIATION)!);
+      this.associationMap.get(associationKey)!.set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.prettyPrintAssociation(portfolio, product)}`);
+    }
+  }
+
+  private static associationMap = new Map<string, Map<Constraints, AssociationOrConstraint>>();
+
+  private static getAssociationKey(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct): string {
+    return hashValues(portfolio.node.addr, product.node.addr, scope.stack.node.addr);
+  }
+
+  private static ensureProductIsAssociatedWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct): string {
+    const associationKey = this.getAssociationKey(scope, portfolio, product);
+    // We can still call this because it will check if the association already exists before creating a new one
+    this.associateProductWithPortfolio(scope, portfolio, product);
+    return associationKey;
+  }
+
+  private static checkConstraintOrAssocationExists(associationKey: string, constraint: Constraints): boolean {
+    if (constraint === Constraints.ASSOCIATION) {
+      return !this.associationMap.has(associationKey);
+    }
+    return !this.associationMap.get(associationKey)!.has(constraint);
```

Comment:
I don't think the `!` in `this.associationMap.get(associationKey)!` is correct... this would crash if you called it before an association would be added. Seems like a very fragile way to write this.

Reply:
I think we had guaranteed elsewhere in code that the association would be added before this path could be called, but this would fail if somehow it was called out of order.   This is gone now though.